src/utils/cuda/cuda_utils.cpp

```diff
@@ -0,0 +1,351 @@
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "cuda_utils.h"
+#include <nixl_log.h>
+
+#ifdef HAVE_CUDA
+
+#include <cuda_runtime.h>
+#include <cuda.h>
+
+#endif
+
+namespace nixlCuda {
+
+/****************************************
+ * CUDA nixlCudaPtr class
+*****************************************/
+
+#ifdef HAVE_CUDA
+
+    class memCtxImpl : public memCtx {
+    private:
+        CUcontext ctx;
+
+        nixl_status_t queryVmm(const void *address, memory_t &type, uint64_t &id);
+        nixl_status_t retainVmmCudaCtx(uint64_t id, CUcontext &newCtx) const;
+        nixl_status_t releaseVmmCudaCtx(uint64_t id) const;
+        nixl_status_t queryCuda(const void *address, memory_t &type, uint64_t &id,
+                                CUcontext &newCtx);
+
+    public:
+
+        memCtxImpl() : memCtx()
+        {    }
+
+        ~memCtxImpl() override {
+            if (MEM_VMM_DEV == memType) {
+                nixl_status_t status = releaseVmmCudaCtx(_devId);
```

Comment:
`enableAddr` called twice may leave unreleased primary device context.
Previously retained context should be released before retaining a new one.

Reply:
`enableAddr` only retains once (`if (MEM_NONE == memType)`).
If the memory type is set to something, `enableAddr` only compares device IDs.
So I'm not sure that I understand the problem you describe here.__

Reply:
I see. I mistook this variable for a local variable of enableAddr, not a class field.